common/locks/priority_semaphore_test.go

```diff
@@ -0,0 +1,247 @@
+		// Wait for other thread to block on this semaphore.
+		s.waitUntilBlockedInSemaphore(1)
+		// Release the semaphore so that blocking thread can resume.
+		semaphore.Release(1)
+	}()
+	<-cLock
+	s.NoError(semaphore.Acquire(context.Background(), PriorityLow, 1))
+}
+
+func (s *prioritySemaphoreSuite) TestTryAcquire_HighAllowedBeforeLow() {
+	semaphore := NewPrioritySemaphore(1)
+	wg := sync.WaitGroup{}
+	s.True(semaphore.TryAcquire(1))
+	wg.Add(1)
+	go func() {
+		s.waitUntilBlockedInSemaphore(2)
+		semaphore.Release(1)
+		wg.Done()
+	}()
+	wg.Add(1)
+	lowAcquired := false
+	go func() {
+		s.NoError(semaphore.Acquire(context.Background(), PriorityLow, 1))
+		lowAcquired = true
+		wg.Done()
+	}()
+	s.NoError(semaphore.Acquire(context.Background(), PriorityHigh, 1))
+	// Checking if LowPriority goroutine is still waiting.
+	s.False(lowAcquired)
+	semaphore.Release(1)
+	wg.Wait()
+	// Checking if LowPriority goroutine acquired semaphore.
+	s.True(lowAcquired)
+}
+
+func (s *prioritySemaphoreSuite) Test_AllThreadsAreWokenUp() {
+	semaphore := NewPrioritySemaphore(10)
+	ctx := context.Background()
+	s.NoError(semaphore.Acquire(ctx, PriorityHigh, 6))
+	s.NoError(semaphore.Acquire(ctx, PriorityHigh, 4))
+
+	wg := sync.WaitGroup{}
+	wg.Add(10)
+	for i := 0; i < 5; i++ {
+		go func() {
+			s.NoError(semaphore.Acquire(ctx, PriorityHigh, 1))
+			wg.Done()
+		}()
+	}
+	for i := 5; i < 10; i++ {
+		go func() {
+			s.NoError(semaphore.Acquire(ctx, PriorityLow, 1))
+			wg.Done()
+		}()
+	}
+
+	// Waiting for all above goroutines to block on semaphore.
+	s.waitUntilBlockedInSemaphore(10)
+
+	semaphore.Release(6)
+	semaphore.Release(4)
+
+	wg.Wait()
+}
+
+func (s *prioritySemaphoreSuite) Test_ContextCanceledBeforeAcquire() {
+	semaphore := NewPrioritySemaphore(1)
+	ctx, cancel := context.WithCancel(context.Background())
+	cancel()
+	s.ErrorIs(semaphore.Acquire(ctx, PriorityHigh, 1), ctx.Err())
+}
+
+func (s *prioritySemaphoreSuite) Test_InvalidPriority() {
+	semaphore := NewPrioritySemaphore(1)
+	s.Panics(func() {
+		_ = semaphore.Acquire(context.Background(), Priority(10), 1)
+	})
+}
+
+func (s *prioritySemaphoreSuite) Test_TimedOutWaitingForLock() {
+	semaphore := NewPrioritySemaphore(1)
+	s.NoError(semaphore.Acquire(context.Background(), PriorityHigh, 1))
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+	s.ErrorIs(semaphore.Acquire(ctx, PriorityHigh, 1), ctx.Err())
+}
+
+func (s *prioritySemaphoreSuite) Test_AcquireMoreThanAvailable() {
+	semaphore := NewPrioritySemaphore(1)
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+	s.ErrorIs(semaphore.Acquire(ctx, PriorityHigh, 2), ctx.Err())
+}
+
+// Checks if n number of threads are blocked in semaphore.
+func (s *prioritySemaphoreSuite) waitUntilBlockedInSemaphore(n int) {
+	s.Eventually(
+		func() bool {
+			buf := make([]byte, 10000)
+			runtime.Stack(buf, true)
+			goroutines := bytes.Split(buf, []byte("\n\n"))
```

Comment:
As I mentioned in my thread on an older revision: this is incorrect usage of `runtime.Stack`, which returns the number of bytes written. This needs to be 
```
buf := make([]byte, 10000)
n := runtime.Stack(buf, true)
goroutines := bytes.Split(buf[:n], []byte("\n\n"))
```

With that change you could also make the 10K alloc above the eventually and reuse it, but it doesn't really matter as its a test /shrug
```